contracts/governance/src/ZeroExVotes.sol

```diff
@@ -0,0 +1,232 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+
+  Copyright 2023 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+pragma solidity ^0.8.17;
+
+import "@openzeppelin/utils/math/SafeCast.sol";
+import "@openzeppelin/utils/math/Math.sol";
+import "@openzeppelin/token/ERC20/ERC20.sol";
+import "@openzeppelin/governance/utils/IVotes.sol";
+import "./IZeroExVotes.sol";
+
+/// @custom:security-contact security@0xproject.com
+contract ZeroExVotes is IZeroExVotes {
+    address public token;
+
+    mapping(address => Checkpoint[]) private _checkpoints;
+    Checkpoint[] private _totalSupplyCheckpoints;
+
+    modifier onlyToken() {
+        require(msg.sender == token, "Only the token is allowed to perform this operation");
+        _;
+    }
+
+    function initialize(address _token) public {
```

Comment:
```suggestion
    function initialize(address _token) public onlyProxy {
```

Reply:
Which proxy do you mean here?

I was assuming that this will only ever be instantiated once during deployment and if we limit it will need to be the deployer themselves but I thought it's not worth the complexity.

Reply:
initializers that can be called on the implementation contract are extremely dangerous. It's best practice to at least add something like

```Solidity
address internal immutable _thisCopy;
constructor () {
  _thisCopy = address(this);
}
modifier onlyProxy() {
  require(address(this) != _thisCopy, "only proxy");
}
function initialize(...) onlyProxy external {
  ...
}
```

I'm pretty sure OZ supplies something equvalent

Reply:
I would still recommend adding `onlyProxy` here to _completely_ prevent initialization of the implementation. It's _probably_ harmless, but let's be conservative.

Reply:
cherry-picked your commit https://github.com/0xProject/protocol/pull/674/commits/fd63a2521ae93e02b7892ae6e2017e0aa5d1da9e